### PR TITLE
Make Thread.VolatileRead and Thread.VolatileWrite obsolete

### PR DESCRIPTION
**Pull request by CollinAlpert**
This PR obsoletes `Thread.VolatileRead` as well as `Thread.VolatileWrite` in favor of `Volatile.Read` and `Volatile.Write` respectively, as per #27997.

I am implementing the fixer for the diagnostic ID in dotnet/roslyn-analyzers#7043. The main reason I want to get this PR merged before completing the fixer, is to confirm `SYSLIB0054` as the diagnostic ID.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs [hunk 1]
```diff
@@ -542,41 +543,93 @@ public void SetCompressedStack(CompressedStack stack)
         public static void MemoryBarrier() => Interlocked.MemoryBarrier();
         public static void Sleep(TimeSpan timeout) => Sleep(WaitHandle.ToTimeoutMilliseconds(timeout));
 
+        [Obsolete(Obsoletions.ThreadVolatileReadMessage, DiagnosticId = Obsoletions.ThreadVolatileReadWriteDiagId)]
```
[the remark this answers is not in the document]
Comment (vcsjones) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs:546]: Attribute changes need to be in the reference definition of the methods as well. https://github.com/dotnet/runtime/blob/5f52977c066a29d124694eb423eff0b866df37a7/src/libraries/System.Threading.Thread/ref/System.Threading.Thread.cs#L132

See other examples of `System.ObsoleteAttribute` in that file.

File: src/libraries/Common/src/System/Obsoletions.cs [hunk 2]
```diff
@@ -171,5 +171,9 @@ internal static class Obsoletions
 
         internal const string AesGcmTagConstructorMessage = "AesGcm should indicate the required tag size for encryption and decryption. Use a constructor that accepts the tag size.";
         internal const string AesGcmTagConstructorDiagId = "SYSLIB0053";
+
+        internal const string ThreadVolatileReadMessage = "Thread.VolatileRead is obsolete. Use Volatile.Read instead.";
```
Comment (vcsjones) [inline src/libraries/Common/src/System/Obsoletions.cs:175]: This needs to be added to https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md

File: src/libraries/System.Threading.Thread/ref/System.Threading.Thread.cs [hunk 3]
```diff
@@ -110,41 +110,67 @@ public sealed partial class Thread : System.Runtime.ConstrainedExecution.Critica
...
         [return: System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute("address")]
         public static object? VolatileRead([System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute("address")] ref object? address) { throw null; }
+        [System.ObsoleteAttribute("Thread.VolatileRead and Thread.VolatileWrite are obsolete. Use Volatile.Read or Volatile.Write respectively instead.", DiagnosticId = "SYSLIB0054", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
         [System.CLSCompliantAttribute(false)]
         public static sbyte VolatileRead(ref sbyte address) { throw null; }
+        [System.ObsoleteAttribute("Thread.VolatileRead and Thread.VolatileWrite are obsolete. Use Volatile.Read or Volatile.Write respectively instead.", DiagnosticId = "SYSLIB0054", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
         public static float VolatileRead(ref float address) { throw null; }
+        [System.ObsoleteAttribute("Thread.VolatileRead and Thread.VolatileWrite are obsolete. Use Volatile.Read or Volatile.Write respectively instead.", DiagnosticId = "SYSLIB0054", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
```
Comment (vcsjones) [inline src/libraries/System.Threading.Thread/ref/System.Threading.Thread.cs:133]: The EditorBrowsableState(Never) also needs to go in to the ref file.

**Review by stephentoub — COMMENTED** [review]
Other than the feedback about EditorBrowsable, LGTM. Thanks.

**Review by buyaa-n — APPROVED** [review]
LGTM Thank you!

**Comment by gewarren** [comment]
Should this have the needs breaking change doc label?

**Comment by stephentoub** [comment]
@jeffhandley, @ericstj, do we treat every [Obsolete] as a breaking change to be documented?

**Comment by ericstj** [comment]
Yes, since it will be source breaking (new warnings - for some as error).  Of course this one is a low impact one since it just requires a rename and a fixer is planned.  If we don't want the breaking change then don't make the APIs obsolete.

**Comment by gewarren** [comment]
The breaking change doc is taken care of with https://github.com/dotnet/docs/pull/40834.

**Pull request merged**